src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -915,21 +906,137 @@ public override void ReportDiagnosticsIfObsolete(Binder binder, SyntaxNode synta
 
             protected override bool TypeSymbolEquals(TypeSymbolWithAnnotations other, TypeCompareKind comparison)
             {
-                var otherLazy = other as LazyNullableType;
-
+                var otherLazy = other as LazyNullableTypeParameter;
                 if ((object)otherLazy != null)
                 {
                     return _underlying.TypeSymbolEquals(otherLazy._underlying, comparison);
                 }
+                return base.TypeSymbolEquals(other, comparison);
+            }
+        }
+
+        /// <summary>
+        /// Non-nullable or null-oblivious type parameter. IsNullable is calculated
+        /// lazily to avoid cycles when binding declarations.
+        /// </summary>
+        private sealed class LazyNonNullableTypeParameter : TypeSymbolWithAnnotations
+        {
+            private readonly TypeParameterSymbol _typeParameter;
+            private readonly bool? _isNullableIfReferenceType;
+
+            internal LazyNonNullableTypeParameter(TypeParameterSymbol typeParameter, bool? isNullableIfReferenceType)
+            {
+                Debug.Assert(isNullableIfReferenceType != true); // Should use LazyNullableTypeParameter.
+                _typeParameter = typeParameter;
+                _isNullableIfReferenceType = isNullableIfReferenceType;
+            }
+
+            public override TypeSymbol TypeSymbol => _typeParameter;
+            public override bool? IsNullable =>  _isNullableIfReferenceType == false || !_typeParameter.IsReferenceType ? (bool?)false : null;
+
+            public override ImmutableArray<CustomModifier> CustomModifiers => ImmutableArray<CustomModifier>.Empty;
+
+            public override TypeSymbolWithAnnotations AsNotNullableReferenceType()
+            {
+                return _isNullableIfReferenceType == false ?
+                    this :
+                    new LazyNonNullableTypeParameter(_typeParameter, isNullableIfReferenceType: false);
+            }
+
+            public override TypeSymbolWithAnnotations AsNullableReferenceType()
+            {
+                return new LazyNullableTypeParameter(_typeParameter.ContainingAssembly, this);
```

Comment:
Random thought: might it make sense to consolidate `LazyNonNullableTypeParameter` and `LazyNullableTypeParameter`?

Reply:
I considered that but the implementations are somewhat distinct. In particular, `LazyUnannotatedTypeParameter` holds on to a `TypeParameterSymbol` directly while `LazyNullableTypeParameter` holds on to the `TypeSymbolWithAnnotations` for the unannotated type parameter.

It might be useful to have a common base class.

---
In reply to: [201558313](https://github.com/dotnet/roslyn/pull/28453#discussion_r201558313) [](ancestors = 201558313)